Sources/Promises/Promise+All.swift

```diff
@@ -133,3 +133,51 @@ public func all<A, B, C>(
   }
   return promise
 }
+
+/// Waits until all of the promises have been fulfilled.
+/// If one of the promises is rejected, then the returned promise is rejected with same error.
+/// If any other arbitrary value or `Error` appears in the array instead of `Promise`,
+/// it's implicitly considered a pre-fulfilled or pre-rejected `Promise` correspondingly.
+/// - parameters:
+///   - queue: A queue to dispatch on.
+///   - promiseA: Promise of type `A`.
+///   - promiseB: Promise of type `B`.
+///   - promiseC: Promise of type `C`.
+///   - promiseD: Promise of type `D`.
+/// - returns: Promise of a tuple containing the values of input promises in the same order.
+public func all<A, B, C, D>(
+  on queue: DispatchQueue = .promises,
+  _ promiseA: Promise<A>,
+  _ promiseB: Promise<B>,
+  _ promiseC: Promise<C>,
+  _ promiseD: Promise<D>
+) -> Promise<(A, B, C, D)> {
+  let promises = [
+    promiseA.objCPromise,
+    promiseB.objCPromise,
+    promiseC.objCPromise,
+    promiseD.objCPromise
+  ]
+  let promise = Promise<(A, B, C, D)>(
+    Promise<(A, B, C, D)>.ObjCPromise<AnyObject>.__onQueue(
+      queue,
+      all: promises
+    ).__onQueue(queue, then: { objCValues in
+      guard let values = objCValues as [AnyObject]?,
+            let valueA = Promise<A>.asValue(values[0]),
+            let valueB = Promise<B>.asValue(values[1]),
+            let valueC = Promise<C>.asValue(values[2]),
+            let valueD = Promise<D>.asValue(values[3])
+      else {
+        preconditionFailure("Cannot convert \(type(of: objCValues)) to \((A, B, C, D).self)")
+      }
+      return (valueA, valueB, valueC, valueD)
+    })
+  )
+  // Keep Swift wrapper alive for chained promises until `ObjCPromise` counterpart is resolved.
+  promises.forEach {
```

Comment:
we could probably get some code re-use here?

Reply:
Should be fine for optimization purposes: no extra func call.